### PR TITLE
Fix static_argnums values leaking into a global jit cache (#16226)

### LINKED ISSUE #16226: `jax.jit(..., static_argnums=...)` keeps strong (not weak) reference in cache, causing memory leak
### Description

```python
import gc
import weakref
import jax

class A:
	pass

def f(x, y):
	return x

f = jax.jit(f, static_argnums=1)
f(1, A)
ref = weakref.ref(A)
del A
gc.collect()
print(ref)  # <weakref at 0x780ddf5062c0; to 'type' at 0x12ff5d0 (A)>
```
Commenting out the `f(1, A)` line instead produces `<weakref at 0x7a4f2bdce360; dead>`.

This use-case turns up in tests, when creating and destroying many JIT'd functions.

### What jax/jaxlib version are you using?

4.11

### Which accelerator(s) are you using?

_No response_

### Additional system info

_No response_

### NVIDIA GPU info

_No response_

### PR DESCRIPTION
**Pull request by kapellirohith**
Fixes #16226

### Root Cause Analysis
The memory leak reported in #16226 originates in the Python-side caching layer, specifically `_process_in_axis_resources` in `pjit.py`. This function is memoized via a global LRU cache whose key includes the statics-carrying avals `FlatTree`. As a result, arbitrary user objects passed via `static_argnums`/`static_argnames` are pinned in memory indefinitely, surviving deletion of both the jitted wrapper and the underlying function — only `jax.clear_caches()` releases them.

Additionally, because resource canonicalization does not depend on static values (statics contribute no leaves to `flatten_axis_resources`, and `in_tree_filtered` already has them filtered out), keying on them caused spurious misses in this canonicalization cache when only static values differed.

### Fix
* Added a `strip_statics()` method to `FlatTree` (`flattree.py`): returns a structural copy with `FTStatic` payloads replaced by a shared `_StrippedStatic` placeholder, preserving lengths and dynamic avals — the cached function's result is unchanged.
* Updated the call site in `pjit.py` to pass `avals_ft.strip_statics()`, severing the lifetime tie between the global cache and user objects.

This addresses the lifetime bug (cache entries outliving the function). The broader weak-vs-strong static keying design question discussed in the issue thread by @mattjj ([comment](https://github.com/jax-ml/jax/issues/16226#issuecomment-1572932916)) and @patrick-kidger ([comment](https://github.com/jax-ml/jax/issues/16226#issuecomment-1572992371)) — i.e. whether caches should ever hold statics weakly based on `__eq__` heuristics — remains a separate conversation; this PR does not change value-based caching semantics while a function is alive.

### Validation & Testing
* **Memory profiling:** RSS over 30 create/destroy cycles of `jit` functions with 8 MB static payloads: **+248 MB linear growth → +25 MB flat warmup**. Static arguments are now freed on `del`.
* **Regression coverage:** Added `test_jit_static_argnums_reference_dropping` to `api_test.py`; it fails without the fix and passes with it.
* **Test suites:** `api_test.py` (844 passed), `pjit_test.py` (796 passed), plus the jit / cache-key / compilation-cache / heap-profiler suites — 1,700+ tests, zero regressions.

### CODE REVIEW

**Comment by dougalm** [comment]
Closing - not worth the code complexity for such a small win.

**Comment by kapellirohith** [comment]
Makes sense. Is the leak itself worth a lighter fix — e.g. only stripping when the static is non-__eq__-comparable — or is the whole direction not worth pursuing
